src/executor/validate.rs

```diff
@@ -0,0 +1,204 @@
+    let column_defs = Rc::new(column_defs);
+    let unique_checks = Rc::new(unique_checks);
+    storage
+        .scan_data(table_name)
+        .await
+        .map(stream::iter)?
+        .try_for_each(|(_, row)| {
+            let column_defs = Rc::clone(&column_defs);
+            let unique_checks = Rc::clone(&unique_checks);
+
+            async move {
+                for check in unique_checks.as_ref() {
+                    let column_index = check.column_index;
+                    let row_val = row.get_value(column_index).unwrap();
+                    if check
+                        .values
+                        .binary_search_by(|v| compare_values_for_sort(v, row_val).unwrap())
+                        .is_ok()
+                    {
+                        return Err(ValueError::DuplicateEntryOnUniqueField(
+                            format!("{:?}", row_val),
+                            column_defs[column_index].name.to_string(),
+                        )
+                        .into());
+                    }
+                }
+                Ok(())
+            }
+        })
+        .await
+}
+
+fn fetch_all_unique_column_indexes(column_defs: &[ColumnDef]) -> Vec<usize> {
+    let mut column_indexes = vec![];
+    for (i, col) in column_defs.iter().enumerate() {
+        if col
+            .options
+            .iter()
+            .any(|opt_def| matches!(opt_def.option, ColumnOption::Unique { .. }))
+        {
+            column_indexes.push(i);
+        }
+    }
+
+    column_indexes
+}
+
+fn fetch_specified_unique_column_indexes(
+    table_column_defs: &[ColumnDef],
+    specified_columns: &[Ident],
+) -> Vec<usize> {
+    let mut column_indexes = vec![];
+    for (i, table_col) in table_column_defs.iter().enumerate() {
+        if table_col
+            .options
+            .iter()
+            .any(|opt_def| match opt_def.option {
+                ColumnOption::Unique { .. } => specified_columns
+                    .iter()
+                    .any(|specified_col| specified_col.value == table_col.name.value),
+                _ => false,
+            })
+        {
+            column_indexes.push(i);
+        }
+    }
+
+    column_indexes
+}
+
+#[derive(Debug)]
+struct UniqueCheck<'a> {
+    column_index: usize,
+    values: Vec<&'a Value>,
+}
+
+fn create_unique_checks<'a, U: 'static + Debug, F>(
+    table_column_defs: &[ColumnDef],
+    unique_column_indexes: &[usize],
+    items: &'a [U],
+    item_to_row: F,
+) -> Result<Vec<UniqueCheck<'a>>>
+where
+    F: Fn(&U) -> &Row,
+{
+    let item_len = items.len();
+    let mut checks = Vec::with_capacity(unique_column_indexes.len());
+    for &column_index in unique_column_indexes {
+        let mut values: Vec<&'a Value> = Vec::with_capacity(item_len);
+        for item in items {
+            let new_val = item_to_row(item).get_value(column_index).unwrap();
+            match values.binary_search_by(|v| compare_values_for_sort(v, new_val).unwrap()) {
```

Comment:
This was impressive, I appreciate.
I agree that this filtering is a good choice.
But how about using sets in `std::collections` rather than using custom implementation?

Reply:
Hi @panarch , I added this function `compare_value_for_sort` as an `internal` (just in this file) implementation of trait `Ord`. Because I need an efficient method to insert and find a Value in the sorted Vector. The function `binary_search/binary_search_by` requires `T` must implement the trait `Ord`, the function `compare_value_for_sort` is used for this.

As you mentioned the sets of `std::collections`, I found [HashSet](https://doc.rust-lang.org/std/collections/struct.HashSet.html) and [BTreeSet](https://doc.rust-lang.org/std/collections/struct.BTreeSet.html). `BTreeSet` needs T must implement the trait `Ord` (and cannot be customed with a closure like `binary_search_by`).

I try to use `HashSet` and add `#[derive(Hash)]` trait to `data::Value`, but the type `f64` cannot been hashed (since `NAN`?). I found [this-discussion](https://internals.rust-lang.org/t/f32-f64-should-implement-hash/5436), it seems that the type `f64` needs to be wrapped by `OrderedFloat` (crate [ordered-float](https://github.com/reem/rust-ordered-float)), but the code seems not good.

What's your option? Do you have any suggestions for this? Thanks.

Reply:
@silathdiir Oh! I was supposed to mention this issue in the first comment but I forgot to write, thanks for the reply.
I also experienced `Eq + Hash` related issue before when I was implementing `GROUP BY` functionality.

Yes you're right. `f64`, floating point number is the problem. `f64` does not implement `Eq`, so there's no way to support `Hash`.
However, it is also reasonable for `f64` not to have `Eq` because of zero comparison issue.
If we want to support hashing for floating point numbers, then the right choice is using something like `Decimal` type, but that is not implemented in GlueSQL now.

What I decided to do was preventing the use of float in the `GROUP BY`.
If someone wants to `GROUP BY` on float fields, then they need to run some functions like `ceil`, `floor` or `round`, or.. they can use `DECIMAL` type. Though we don't have all that features, but we will have.

I also agree that using `OrderedFloat` is not a good idea. I think we don't need to use something.. tricks, we are making a new sql database, it is not copying other db's functionalities, so..


fyi.
This might be useful to see.

https://github.com/gluesql/gluesql/blob/036080b7c5155a2bb6b8a26bdaa6a26805019e53/src/executor/aggregate/hash.rs#L7

https://github.com/gluesql/gluesql/blob/036080b7c5155a2bb6b8a26bdaa6a26805019e53/src/data/value.rs#L195

I added some `GroupKey` related codes to `data/value.rs`,
I wanted to gather codes which handle the types in `Value` enum to `data/value.rs`.
Then, when we consider to add any new data type to `Value`, it would be easy to work on it because we don't need to take care of any other codes outside of `data/value.rs`. That was my intention.

---

Ok, let's make the issue simple.
Doing same approach with `GROUP BY`,
* Float type cannot have `UNIQUE` constraint

Then you may need to add a new error for `CREATE` query which is normal error type.
Then, in the `validate.rs`, one `Conflict-` prefixed error will be required for floating point + unique validation case.
How do you think of this?